### DISCUSSION TITLE
Morpheus GraphQL Named resolvers

### DISCUSSION THREAD

**Discussion by nalchevanidze** [brandonchinn178/brandonchinn178.github.io#16] [General]
Hi, i saw your [post](https://brandonchinn178.github.io/blog/2021/08/04/graphql-server-haskell.html) and as a Response i implemented,

https://github.com/morpheusgraphql/morpheus-graphql/pull/628/files

you can check out it and would like to hear your view about it.

**Comment by brandonchinn178** [comment]
Hi! Thanks for reading my blog post! I'm a bit unclear what that PR does, both as a whole and what aspect of the blog post it addresses

**Comment by nalchevanidze** [comment]
@brandonchinn178 i addressed this particular problem:

> After going through this list, it seems like morpheus-graphql would do an okay job, but it would require redesigning my project from being organized by domain to being organized by tech layer (e.g. database layer > business logic layer > graphql layer). I rather like having my project organized by domain, so that the full end-to-end workflow for a particular workflow (e.g. get a user) is colocated.

this [post](https://morpheusgraphql.com/named-resolvers) is actually quite poorly documented, but it shows how to merge two domain-organised feature-apps into a single GraphQL app. we define `Posts` and `Authors` in separate (folder) file, and overload type `Post` with additional field `author`. and since type `Post` in `Authors.hs` and in `Posts.hs` have same dependency, interpreter can safely merge them. the feature will be introduced in `v0.18.0` and  is still experimental, but will be evolve gradually. 

furthermore version 0.18.0 will also support code-gen based on `CLI`. 

anyway thanks for your post about problems with morpheus GraphQL, they will be addressed.

**Reply by brandonchinn178** [reply]
Oh, super cool! Yes, that looks great! I assume `ID` would be defined in a separate file that both `Posts.hs` and `Authors.hs` import? Shouldn't `Author` and `Post` have different `ID` types?

There seem to be two remaining issues with this: making sure the combined App has all the resolvers implemented, and autogenerating the resolver types from the schema DSL. I guess if you autogenerate types per `.graphql` schema file, then `authors.graphql` can generate `Authors/Generated.hs` with `author` being the only field in the `Post` resolver.

**Reply by nalchevanidze** [reply]
> Shouldn't Author and Post have different ID types?

authors and users can have different dependency type. however if type `User` defined in authors and users have different dependency type merging should fail.

>  making sure the combined App has all the resolvers implemented

that is not a problem, since deriving will fail if you forgot to implement resolver for all type.

> and autogenerating the resolver types from the schema DSL

that is true, it will be tricky.

**Reply by brandonchinn178** [reply]
Oh I see. How does `authorsApp <> postsApp` know that the two `Query` data types are the same? Similarly, how does it know the two `Post` types are the same, and the fields should be merged?

**Reply by nalchevanidze** [reply]
if these apps have type with same name like "Post" it will assume that they are describing same entity and will try merge them.
merging will check if:
    - both have same kind (object, unions ... )
    - both have same dependency ( where dependency will be checked by uniq identifier [fingeprint](https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-Fingerprint-Type.html#t:Fingerprint). this step is necessary to ensure safe argument encoding and decoding for resolver functions.  
  if all criterieas are fullfilled then app will merge these types.
